{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #6_ 3D visualization and mutagenesis of protein Colab_hgprt.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyN95crEroXUp+Xd4K/ZMB3F"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422–1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"R8qI2W99X6eq"},"source":["# Examining the 3D structure of your protein"]},{"cell_type":"markdown","metadata":{"id":"ooKhzn30nT8g"},"source":["\n","In this section, you will search the Protein Data Bank for the 3D structure of the HPRT protein. You will use the name 1BZY as your search query. For this, you will use a package called prody."]},{"cell_type":"markdown","metadata":{"id":"p0mkTQbjqkS-"},"source":["## Install and import the necessary packages:\n","\n","\n","**ProDy** is a free and open-source Python package for protein structural dynamics analysis.\n","  import * allows for use of the proteins module within ProDy\n","\n","**Py3Dmol** is a widget used to embed an interactive viewer in a notebook.\n","\n","REF:\n","* http://prody.csb.pitt.edu/\n","* https://pypi.org/project/py3Dmol/\n","* Nicholas Rego and David Koes\n","3Dmol.js: molecular visualization with WebGL\n","Bioinformatics (2015) 31 (8): 1322-1324 doi:10.1093/bioinformatics/btu829\n","\n","\n"]},{"cell_type":"code","metadata":{"id":"ombRngjw9mmA"},"source":["!pip install prody\n","from prody import *\n","!pip install py3Dmol\n","import py3Dmol"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3XY-nYDUsPpg"},"source":["STEP #1: Download the PDB file of your protein. You will use it later on for visualization purposes."]},{"cell_type":"code","metadata":{"id":"hqZq0zvl97aM"},"source":["# Download the PDB file by writing the 4-letter code inside the parenthesis and running the cell\n","fetchPDB('####')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ps_hPTTwtO-_"},"source":["STEP #2: Obtain information about the crystal structure of your protein by parsing the data in the PDB file.\n","\n","You can use the header attribute to any of the words below to get more information.\n","\n","* 'A'\n","* 'related_entries'\n","* 'sheet'\n","* 'classification'\n","* 'reference'\n","* 'title'\n","* 'sheet_range'\n","* 'polymers'\n","* 'resolution'\n","* 'space_group'\n","* 'helix_range'\n","* 'chemicals'\n","* 'experiment'\n","* 'helix'\n","* 'version'\n","* 'authors'\n","* 'identifier'\n","* 'deposition_date'\n","* 'biomoltrans'"]},{"cell_type":"code","metadata":{"id":"2Sqo9CqE-VAy"},"source":["# Let's parse the PDB file for information\n","\n","# Write the 4-leter code within the parenthesis and set 'header' to True\n","# 'header' attribute will let us obtain more information\n","\n","atoms, header = parsePDB('####', header=####)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"zPKf-Ii2u3EP"},"source":["# Obtain the reference of the crystal structure, the article it came from\n","print(header['reference'])\n","\n","# Obtain which experiment produced the crystal structure by writting 'experiment' in the brakets\n","print(header['####'])\n","\n","# Obtain the resolution of the crystal structure\n","# YOUR CODE HERE"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dSmqj1FSymp0"},"source":["STEP #3: Read the abstract of the article you obtained in the previous cells. For this, you will search another NCBI database called PubMed and extract the abstract of the article."]},{"cell_type":"code","metadata":{"id":"DxjiYVCKyhTP"},"source":["from Bio import Entrez # Here we need Entrez once more\n","# Searching for the article abstract in the database Pubmed\n","Entrez.email = 'YOUR EMAIL HERE'\n","\n","# Write the id, rettype = abstract and retmode, text\n","# db = database, in this case we are using the PubMed database\n","query6 = Entrez.efetch(db='####', id = '####', rettype = '####', retmode = '####')\n","\n","# Reading the query and printing it\n","print(query6.read())\n","\n","# Closing the query\n","query6.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"v9L-sqQMY9d5"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"TWsamYk0VgtW"},"source":["\n","1. What is the resolution of this crystal structure and what does “resolution”\n","mean for a crystal structure?"]},{"cell_type":"markdown","metadata":{"id":"R7-y1q2OVjSI"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"aDAhtnwL3sqN"},"source":["2. Go to https://www.rcsb.org/ and access the entry for this protein.\n","  What organism was H-Ras obtained from for the crystal structure?"]},{"cell_type":"markdown","metadata":{"id":"fWgfYVsGVmbz"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"HCwizhyuZ0MU"},"source":["## Examining Secondary Structure\n","Now let's look at the protein in 3D!"]},{"cell_type":"code","metadata":{"id":"BWYLcJBb89b2"},"source":["# Initialize the viewer with the appropriate PDB structure\n","\n","# Write the 4-letter code of your protein inside the parenthesis: 'pdb:####'\n","view = py3Dmol.view(query = 'pdb:####')\n","\n","# Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","# Here we set the visualization style for the protein\n","view.setStyle({'cartoon': {'color':'skyblue'}})\n","\n","# Here we set the visualization style for the ligand\n","view.addStyle({'resi':167},{'stick':{'colorscheme':'grayCarbon'}})\n","\n","view.zoomTo()\n","view.render()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KQWpajpzMC6t"},"source":["###Available colors to choose from###\n","* https://www.w3schools.com/cssref/css_colors.asp\n","* Use the name in lowercase letters.\n","\n","\n","##You can also set specific color schemes##\n","Try a few to see what they look like!\n","* ssPyMOL\t- PyMol secondary colorscheme\n","* ssJmol - Jmol secondary colorscheme\n","* Jmol - Jmol primary colorscheme\n","* default\t- default colorscheme\n","* amino\t- amino acid colorscheme\n","* shapely\t- shapely protein colorscheme\n","* nucleic\t- nucleic acid colorscheme\n","* chain\t- standard chain colorscheme\n","* gradient - Allows the user to provide a gradient to the colorscheme. Is either a $3Dmol.Gradient object or the name of a built-in gradient (rwb, roygb, sinebow)\n","\n","Code:\n","```\n","viewer.setStyle({'cartoon':{'colorscheme':'amino'}})\n","\n"]},{"cell_type":"markdown","metadata":{"id":"oH1zStGroqbE"},"source":["## Make your own representation of the protein! \n","\n","* Change the color and style of the protein and ligand by replicating the code from previous cells.\n","* Write and execute your code below.\n","* Take a screenshot of a good view of your protein for your final report\n"]},{"cell_type":"code","metadata":{"id":"EMcTbIzOrPMe"},"source":["# YOUR CODE HERE!"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tP7UuO5JRKZ0"},"source":["## Compare the secondary structure to the predicted secondary structure provided by PSIPRED##\n","\n","PSI-blast based secondary structure PREDiction (PSIPRED) is a method used to investigate protein structure. It uses artificial neural network machine learning methods in its algorithm.\n","\n","REF:\n","* https://en.wikipedia.org/wiki/PSIPRED\n","\n"]},{"cell_type":"markdown","metadata":{"id":"-lKjrLxvRxp3"},"source":["###Load the PSIPRED file with the predicted structure by running the cell below.###\n","\n","Note the 'Pred' line indicates whether the amino acid (AA) is predicted to be in a coil (C), strand (E), or alpha-helix (H)."]},{"cell_type":"code","metadata":{"id":"eec-R_k_5rlo"},"source":["# Load PSIPRED file with the predicted structure\n","\n","from IPython.display import IFrame\n","IFrame('https://drive.google.com/file/d/1tIgBuLMpwWQKpxjsnnuxF2ij6-JlFgis/preview', width=600, height=300)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Rpb9sFsS3zZY"},"source":["## Answer the following questions:##\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"RJA5GrK5TSeO"},"source":["1. Carefully examine the secondary structure in the crystal structure and\n","record any positions where the PSIPRED predictions were incorrect."]},{"cell_type":"markdown","metadata":{"id":"mLpgdjeLTXWV"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"APFz8dp6TTPn"},"source":["2. PSIPRED states its predictions are ~80% correct. Do you agree this is a\n","good estimate of the accuracy? Explain."]},{"cell_type":"markdown","metadata":{"id":"HOrMzTKJTY9x"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"vMrdzCd1QOWA"},"source":["# Modeling the mutation\n","\n","Your patient has a mutation in the gene that codes for this protein. In this section you will observe how the mutation affects the function of the protein by observing where the mutation is located and performing a mutation back to the original amino acid residue. "]},{"cell_type":"markdown","metadata":{"id":"rMay7Mws2y7P"},"source":["## Install and import the necessary packages:\n","\n","\n","**Pymol** is an open source molecular visualization system. Like **Py3Dmol**, it allows for 3D visualization of proteins. In this module, we will use it to perform the amino acid mutagenesis. \n","\n","REF:\n","* https://pymol.org/2/\n","* https://colab.research.google.com/github/pb3lab/ibm3202/blob/master/tutorials/lab02_molviz.ipynb#scrollTo=0M8qd0sTyb9p\n","* http://3dmol.csb.pitt.edu/doc/types.html#AtomSelectionSpec"]},{"cell_type":"code","metadata":{"id":"wEwABuh_QQ0z"},"source":["# The code below installs pymol\n","! apt-get install pymol "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"i9QcWOuw6-c5"},"source":["STEP #1: Select the residue you are going to mutate and set it to stick representation.\n"]},{"cell_type":"code","metadata":{"id":"8t01-oeH69Bf"},"source":["# Initialize the viewer with the appropriate PDB structure\n","# Write the 4-letter code of your protein inside the parenthesis: 'pdb:####'\n","view = py3Dmol.view(query = 'pdb:####')\n","\n","# Here we set the background color\n","view.setBackgroundColor('####')\n","\n","# Here we set the visualization style for the protein\n","view.setStyle({'cartoon': {'color':'####'}})\n","\n","# Here we set the visualization style for the ligands (Magnesium, phosphoric acid, pyrophosphate)\n","view.addStyle({'resi':900},{'stick':{'colorscheme':'####'}})\n","view.addStyle({'resi':300},{'stick':{'colorscheme':'####'}})\n","view.addStyle({'resi':400},{'stick':{'colorscheme':'####'}})\n","\n","# Here we set the visualization style for the Aspartate residue\n","# Write '193' after 'resi'\n","\n","view.addStyle({'resi':###},{'stick':{'colorscheme':'####'}})\n","\n","# Label the residue! \n","view.addResLabels({'resi':['###']})\n","\n","view.zoomTo()\n","view.render()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zAnX0spi-btF"},"source":["STEP #2: Perform the mutation! "]},{"cell_type":"code","metadata":{"id":"yICMEQAz5Ql3"},"source":["# Write and execute the command file that mutates the residue\n","\n","# Open the file you will write the new structure to\n","\n","# Write 'NAME OF FILE.pml' before the 'w'\n","com_file = open('NAME OF FILE HERE.pml','w')\n","\n","# Write the file with the following commands:\n","com_file.write('''\n","\n","# Load the PDB file. Write the pdb code in the #### space provided\n","load ####.pdb.gz \n","\n","# Specify the program (wizard) to run within pyMol. Write 'mutagenesis' in the parenthesis.\n","cmd.wizard(\"####\")\n","cmd.do(\"refresh_wizard\")\n","\n","# Specify the reside to mutate and which residue to mutate it to\n","residue_to_change = \"##/\" # Which residue are you changing?\n","residue_mutated = \"###\" # What residue are you changing it to? (write the 3 letter amino acid code in CAPS)\n","\n","for i in dir(cmd.get_wizard()): print i\n","cmd.get_wizard().set_mode(residue_mutated)\n","cmd.get_wizard().do_select(residue_to_change)\n","\n","# Select the rotamer (conformation of residue (direction) this can be set to 1)\n","cmd.frame(1)\n","\n","# Apply the mutation\n","cmd.get_wizard().apply()\n","\n","# Save the new structure as a new file\n","save NAME THIS FILE.pdb\n","''')\n","\n","# Close the command file\n","com_file.close()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"-YpK_Ue3T_aU"},"source":["# Execute the command file created in the previous cell\n","!pymol -c NAME OF FILE.pml"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"NY49eeY8TRBa"},"source":["STEP #3: Visualize the protein again"]},{"cell_type":"code","metadata":{"id":"rYgYzbjqM-Lk"},"source":["# Initialize the viewer with the appropriate PDB structure\n","\n","view=py3Dmol.view()\n","\n","# Read ('r') the file you created in the previous step\n","view.addModel(open('FILE NAME HERE.pdb', 'r').read(),'pdb')\n","\n","# Here we set the background color\n","view.setBackgroundColor('###')\n","\n","# Here we set the visualization style for the protein\n","view.setStyle({'cartoon': {'color':'###'}})\n","\n","# Here we set the visualization style for the ligands (Magnesium, phosphoric acid, pyrophosphate)\n","view.addStyle({'resi':900},{'stick':{'colorscheme':'####'}})\n","view.addStyle({'resi':300},{'stick':{'colorscheme':'####'}})\n","view.addStyle({'resi':400},{'stick':{'colorscheme':'####'}})\n","\n","# Here we set the visualization style for the Aspartate residue\n","# Write '193' after 'resi'\n","\n","view.addStyle({'resi':###},{'stick':{'colorscheme':'###'}})\n","\n","# Label the residue! \n","view.addResLabels({'resi':['###']})\n","\n","\n","view.zoomTo()\n","view.render()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Take a screenshot of a good view/angle of the active site."],"metadata":{"id":"QF2bjbrZ0rep"}},{"cell_type":"markdown","source":["## Answer the following questions:##\n","Input your answer in the cell below each question and press SHIFT+ENTER."],"metadata":{"id":"6AUMmvf3xUAw"}},{"cell_type":"markdown","source":["1. Compare the active site before and after the mutation. After the visual inspection. do you think the mutated amino acid closer or farther from the ligand? "],"metadata":{"id":"EGoWhW9F0tTQ"}},{"cell_type":"markdown","source":["Answer here"],"metadata":{"id":"4w4R3zAS1Oap"}},{"cell_type":"markdown","source":["2. Write your hypothesis explaining how the mutation affects the protein function."],"metadata":{"id":"Nr0NQHIL1Q6U"}},{"cell_type":"markdown","source":["Answer here"],"metadata":{"id":"dAaVeciT1YJX"}}]}