{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #7_ OMIM Colab_sod.ipynb","provenance":[],"collapsed_sections":[],"mount_file_id":"183v7bTGse6nrZM2wak57V9_ddZ46qY3B","authorship_tag":"ABX9TyP9NkMfaHi4p7sTZOKC49cI"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422–1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"yD8hKEUk3Ot4"},"source":["# Investigating the biological impact of the mutation and its possible role in human disease\n","For this section, your research will focus on investigating the biological impact of the mutation you are studying. To do this, you will use the OMIM and KEGG databases."]},{"cell_type":"markdown","metadata":{"id":"Pp2KQ7uEQJkm"},"source":["## OMIM Search for information on genetic diseases\n","\n","The **OMIM** (Online Mendelian Inheritance of Man) database contains short, referenced reviews about genetic loci and genetic diseases. It\n","can be a very useful resource for finding out what type of research has been done on a gene or a disease.\n","\n","REF:\n","* http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM"]},{"cell_type":"markdown","metadata":{"id":"LMm1YA03QaCu"},"source":["## Install and import the necessary packages:\n","\n","The **romim package** was created to query the OMIM database but it runs in R. \n","\n","**R** is another programming language so you will need to install **rpy2** to run R code in Google Colab.\n","\n","**Methods** and **remotes** are R packages that help us both install the package and use the functions in the code.\n","\n","The **XML** package will be used to read the results that you obtain from your database searches. \n","\n","REF:\n","* https://github.com/davetang/romim"]},{"cell_type":"code","metadata":{"id":"yX04cBc0tuEv"},"source":["# Install the rpy2 interface to run R code\n","%load_ext rpy2.ipython"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"cqq6keGK0ynj"},"source":["%%R # This must precede all R code in Colab, to allow R code to run \n","\n","# Installing the main package\n","# Note how different it is from Python code\n","remotes::install_github('davetang/romim')\n","\n","# Import the library associated with the package\n","library(romim)\n","\n","# Intalling several packages\n","install.packages('XML')\n","install.packages('methods')\n","install.packages(\"remotes\")\n","\n","# Press 1 and ENTER when prompted"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ilBRqKQGMlWH"},"source":["## Obtaining the ID number (called mim number) associated with the SOD1 gene entry in OMIM"]},{"cell_type":"code","metadata":{"id":"4hQQALHDGlVw"},"source":["%%R # This must precede all R code in Colab, to allow R code to run \n","\n","# To access OMIM, we will use this key which will work as our password to access the database\n","set_key('4PUvWRqSSD2BuprIVAP_VQ') \n","\n","# Write SOD1 in the parenthesis below to obtain the entry for SOD1\n","gene_to_omim('###', show_query=TRUE)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, the link above is a file with our information. Let's download and parse it."],"metadata":{"id":"t04n8P9yl6ZB"}},{"cell_type":"code","source":["%%R\n","# Saving the file:\n","\n","# Paste the url between the quotes below\n","url <- '#####'\n","\n","\n","# Set destination (write the file path and a file name followed by .xml)\n","destfile <- '/content/drive/MyDrive/Colab_Notebooks/sod_project_files/FILE NAME HERE.xml'\n","\n","# Download the file\n","download.file(url, destfile)\n"],"metadata":{"id":"UhastgNIkp7A"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Parse the file for readability\n","# Find the file in your Drive and copy and paste the path where it says WHERE IS THE FILE? Leave the quotation marks!\n","\n","result <- xmlParse(file = 'WHERE IS THE FILE?')\n","\n","# Read the file\n","read <- read_xml('WHERE IS THE FILE?' )\n","\n","# Find the mim number\n","num <- xml_find_all(read, \".//mimNumber\")\n","\n","# Display the number\n","num"],"metadata":{"id":"sBWzAFDkqSrd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JO11cwOuNQzA"},"source":["Write down the mim number for the entry from the results above."]},{"cell_type":"markdown","metadata":{"id":"LCTL91M9NfUx"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"D9YzHQGWberR"},"source":["## Using OMIM to obtain more information about the gene\n","This time you will search the OMIM 'SOD1' entry for information.\n","\n","The function 'get_omim' helps you do just that since we can set certain arguments to 'TRUE' and obtain specific information about the entry.\n","\n","Run the next cell to see a list of Arguments that you can access."]},{"cell_type":"code","metadata":{"id":"eib4aT-cqMRo"},"source":["%%R\n","help(get_omim)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Obtain information from the entry"],"metadata":{"id":"baghE8ey219O"}},{"cell_type":"code","source":["# Search OMIM again to obtain description information\n","%%R\n","set_key('KEY HERE')\n","\n","# Using mim number to get the entry, set 'text' argument to true\n","omim_result <- get_omim(###, text = ###)\n","\n","# Save the xml to a file \n","saveXML(omim_result, file='FILE NAME HERE.xml')"],"metadata":{"id":"9MzZcfme1cnB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data4 = open('refdata4.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data4)\n","allele_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//textSection'):\n","    allele = []\n","    ref_list = []\n","\n","    if count == 0:\n","      des = member.find('.//textSectionTitle').tag\n","      allele_head.append(des)\n","      \n","      mut = member.find('.//textSectionContent').tag\n","      allele_head.append(mut)\n","     \n","      csvwriter.writerow(allele_head)\n","      count = count + 1\n","       \n","    des = member.find('.//textSectionTitle').text\n","    allele.append(des)\n","    \n","    mut = member.find('.//textSectionContent').text\n","    allele.append(mut)\n","\n","    csvwriter.writerow(allele)\n","  \n","Ref_data4.close()\n","\n","data4= pd.read_csv(\"refdata4.csv\")\n","pd.set_option('display.max_rows', None)\n","pd.set_option('display.max_columns', None)\n","pd.set_option('display.width', None)\n","pd.set_option('display.max_colwidth', -1)\n","\n","data4"],"metadata":{"id":"B9CNi81P1v-c","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"J0XqkjVAeUa9"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","source":["Read the section titled **Description**."],"metadata":{"id":"Rjwo6ak_yUWN"}},{"cell_type":"markdown","metadata":{"id":"iTIm0QiSef5J"},"source":["1. What metal does SOD2 contain?  \n"]},{"cell_type":"markdown","metadata":{"id":"asWjhVSbekgV"},"source":["Answer here"]},{"cell_type":"markdown","source":["2. Where is SOD2 found in cells?\n","\n"],"metadata":{"id":"PMkMlTJiyWn_"}},{"cell_type":"markdown","source":["Answer here"],"metadata":{"id":"26EwstjbyaJT"}},{"cell_type":"markdown","source":["Within the section **Animal Model**, search for the paragraph about the H46R mutation (Hint:  look for “Liu et al.” at the start of the paragraph.) You can use CTRL+F on your machine."],"metadata":{"id":"Q909S-LMTaTa"}},{"cell_type":"markdown","source":["3. What did Liu et al. (2000) find out about the copper binding site in the H46R mutant?\n"],"metadata":{"id":"VzWoPdvBTnzX"}},{"cell_type":"markdown","source":["Answer here\n"],"metadata":{"id":"JEyojW2NTpAK"}},{"cell_type":"markdown","metadata":{"id":"-jjAPZqz0rKB"},"source":["### Now read information about allelic variants \n","An allele is a variant of a gene were the DNA sequence differs between two or more variants. \n","\n","Allelic variation describes the presence or number of different allele forms at a particular locus (locus or loci = place) on a chromosome.\n","\n","REF:  \n","* https://warwick.ac.uk/fac/sci/lifesci/research/vegin/geneticimprovement/diversitycollection/allelicvariation/\n"]},{"cell_type":"code","metadata":{"id":"xJqhcDReyBPC"},"source":["# Search OMIM again but with SOD2 mim Number to obtain description info\n","%%R\n","set_key('KEY HERE')\n","\n","omim_result <- get_omim(###, allelicVariantList = ###)\n","\n","saveXML(omim_result, file='NAME OF FILE.xml')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2Wm9udQGd6If"},"source":["### Display the results in the form of a table"]},{"cell_type":"code","metadata":{"id":"Fos9xqDVN1KQ","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data4 = open('refdata4.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data4)\n","allele_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//allelicVariant'):\n","    allele = []\n","    ref_list = []\n","\n","    if count == 0:\n","      des = member.find('.//mutations').tag\n","      allele_head.append(des)\n","      \n","      mut = member.find('.//text').tag\n","      allele_head.append(mut)\n","     \n","      csvwriter.writerow(allele_head)\n","      count = count + 1\n","       \n","    des = member.find('.//mutations').text\n","    allele.append(des)\n","    \n","    mut = member.find('.//text').text\n","    allele.append(mut)\n","\n","    csvwriter.writerow(allele)\n","  \n","Ref_data4.close()\n","\n","data4= pd.read_csv(\"refdata4.csv\")\n","pd.set_option('display.max_colwidth',10000)\n","\n","data4\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MmIORdPrqApf"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"QK7nCYnCqTz-"},"source":["4. How does the mean survival after disease onset differ for patients with ALS who have the H46R SOD1 mutations from patients with other types of mutations?\n","\n"]},{"cell_type":"markdown","metadata":{"id":"7VkeJAvZqTz-"},"source":["Answer here"]},{"cell_type":"markdown","source":["5. What did the Aoiki et al. 1993 article report about why the H46R SOD1 enzyme is less active?"],"metadata":{"id":"wRJ8EczuqlvZ"}},{"cell_type":"markdown","source":["Answer here"],"metadata":{"id":"ZdNWUBTVqqZF"}},{"cell_type":"markdown","source":["## Now search for information about ALS in the OMIM database"],"metadata":{"id":"pOK447yJT5-x"}},{"cell_type":"code","source":["%%R # This must precede all R code in Colab, to allow R code to run \n","\n","# To access OMIM, we will use this key which will work as our password to access the database\n","set_key('KEY HERE') \n","\n","# First lets get a list of the entries associated with SOD2\n","# Write SOD2 in the parenthesis below to create a list of the entries\n","my_list <- gene_to_omim('####')\n","\n","# Now lets obtain the mim number (with get_omim) and list it with our entries\n","my_list_omim <- sapply(my_list, get_omim)\n","\n","# This will append the title of the entry to the list\n","sapply(my_list_omim, get_title)"],"metadata":{"id":"TI7yBvmg5PIp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ToDePBtz6Qit"},"source":["Write down the ID number for 'AMYOTROPHIC LATERAL SCLEROSIS' from the results above. (ID precedes the entry title)"]},{"cell_type":"markdown","metadata":{"id":"FfQim0uU6Qiz"},"source":["Answer here"]},{"cell_type":"markdown","source":["## Using OMIM to obtain more information about the disease\n","This time you will search the OMIM 'AMYOTROPHIC LATERAL SCLEROSIS' entry for information."],"metadata":{"id":"o5f9w4Fb6nZS"}},{"cell_type":"markdown","source":["### Start by setting 'text' to TRUE"],"metadata":{"id":"XN9WupPp74Af"}},{"cell_type":"code","source":["%%R\n","set_key('####') # The key must be added before every request\n","\n","# Using mim number to get the article list\n","# Write the mim number inside the parenthesis and set 'text' to 'TRUE'\n","omim_result <- get_omim(### , ### = ###)\n","\n","# Save the results as an XML file\n","saveXML(omim_result, file='NAME OF FILE.xml') # Write a file name\n","\n","# File name will display was output for this cell\n"],"metadata":{"id":"IZ0-nDwt7pN_"},"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"qkIEx-fF_reo","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data5 = open('refdata4.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data5)\n","allele_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//textSection'):\n","    allele = []\n","    ref_list = []\n","\n","    if count == 0:\n","      des = member.find('.//textSectionTitle').tag\n","      allele_head.append(des)\n","      \n","      mut = member.find('.//textSectionContent').tag\n","      allele_head.append(mut)\n","     \n","      csvwriter.writerow(allele_head)\n","      count = count + 1\n","       \n","    des = member.find('.//textSectionTitle').text\n","    allele.append(des)\n","    \n","    mut = member.find('.//textSectionContent').text\n","    allele.append(mut)\n","\n","    csvwriter.writerow(allele)\n","  \n","Ref_data5.close()\n","\n","data5= pd.read_csv(\"refdata4.csv\")\n","pd.set_option('display.max_rows', None)\n","pd.set_option('display.max_columns', None)\n","pd.set_option('display.width', None)\n","pd.set_option('display.max_colwidth', -1)\n","\n","data5"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"M_prgY7epGPF"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","source":["Read the section titled **Clinical features**."],"metadata":{"id":"kG6uTDlqpGPF"}},{"cell_type":"markdown","metadata":{"id":"evV74SdDpGPF"},"source":["6. What distinguishes Type 1, Type 2, and Type 3 forms of ALS?\n"]},{"cell_type":"markdown","metadata":{"id":"0mQoUiivpGPF"},"source":["Answer here"]}]}