{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #7: OMIM Colab.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyMRKHC1NdcKGi19FDha8R67"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422â€“1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"yD8hKEUk3Ot4"},"source":["# Investigating the biological impact of the mutation and its possible role in human disease\n","For this section, your research will focus on investigating the biological impact of the mutation you are studying. To do this, you will use the OMIM and KEGG databases."]},{"cell_type":"markdown","metadata":{"id":"Pp2KQ7uEQJkm"},"source":["## OMIM Search for information on genetic diseases\n","\n","The **OMIM** (Online Mendelian Inheritance of Man) database contains short, referenced reviews about genetic loci and genetic diseases. It\n","can be a very useful resource for finding out what type of research has been done on a gene or a disease.\n","\n","REF:\n","* http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM"]},{"cell_type":"markdown","metadata":{"id":"LMm1YA03QaCu"},"source":["## Install and import the necessary packages:\n","\n","The **romim package** was created to query the OMIM database but it runs in R. \n","\n","**R** is another programming language so you will need to install **rpy2** to run R code in Google Colab.\n","\n","**Methods** and **remotes** are R packages that help us both install the package and use the functions in the code.\n","\n","The **XML** package will be used to read the results that you obtain from your database searches. \n","\n","REF:\n","* https://github.com/davetang/romim"]},{"cell_type":"code","metadata":{"id":"yX04cBc0tuEv"},"source":["# Install the rpy2 interface to run R code\n","%load_ext rpy2.ipython"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"cqq6keGK0ynj"},"source":["%%R # This must precede all R code in Colab, to allow R code to run \n","\n","# Installing the main package\n","# Note how different it is from Python code\n","remotes::install_github('davetang/romim')\n","\n","# Import the library associated with the package\n","library(romim)\n","\n","# Intalling several packages\n","install.packages('XML')\n","install.packages('methods')\n","install.packages(\"remotes\")\n","\n","# Press 1 and ENTER when prompted"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ilBRqKQGMlWH"},"source":["## Obtaining the ID number (called mim number) associated with the lung cancer entry in OMIM"]},{"cell_type":"code","metadata":{"id":"4hQQALHDGlVw"},"source":["%%R # This must precede all R code in Colab, to allow R code to run \n","\n","# To access OMIM, we will use this key which will work as our password to access the database\n","set_key('4PUvWRqSSD2BuprIVAP_VQ') \n","\n","# First lets get a list of the entries associated with KRAS2\n","# Write KRAS2 in the parenthesis below to create a list of the entries\n","\n","my_list <- gene_to_omim('####')\n","\n","# Now lets obtain the mim number (with get_omim) and list it with our entries\n","my_list_omim <- sapply(my_list, get_omim)\n","\n","# This will append the title of the entry to the list\n","sapply(my_list_omim, get_title)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JO11cwOuNQzA"},"source":["Write down the ID number for 'LUNG CANCER' from the results above. (ID precedes the entry title)"]},{"cell_type":"markdown","metadata":{"id":"LCTL91M9NfUx"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"D9YzHQGWberR"},"source":["## Using OMIM to obtain more information about the disease\n","This time you will search the OMIM 'LUNG CANCER' entry for information.\n","\n","The function 'get_omim' helps you do just that since we can set certain arguments to 'TRUE' and obtain specific information about the entry.\n","\n","Run the next cell to see a list of Arguments that you can access."]},{"cell_type":"code","metadata":{"id":"eib4aT-cqMRo"},"source":["%%R\n","help(get_omim)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"qUJvtB-vUREd"},"source":["### Start by setting 'referenceList' to TRUE\n"]},{"cell_type":"code","metadata":{"id":"HMjwV3ia2gKV"},"source":["%%R\n","set_key('4PUvWRqSSD2BuprIVAP_VQ') # The key must be added before every request\n","\n","# Using mim number to get the article list\n","# Write the mim number inside the parenthesis and set 'referenceList' to 'TRUE'\n","omim_result <- get_omim(####, referenceList = ####)\n","\n","# Save the results as an XML file\n","saveXML(omim_result, file='FILE NAME HERE.xml') # Write a file name\n","\n","# File name will display as output for this cell\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"kGLxgX1JVLne"},"source":["### Display the results in the form of a table"]},{"cell_type":"code","metadata":{"id":"qkIEx-fF_reo","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = '' #@param {type:\"string\"}\n","\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data = open('refdata.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data)\n","authors_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//reference'):\n","    authors = []\n","    ref_list = []\n","    if count == 0:\n","        author = member.find('.//authors').tag\n","        authors_head.append(author)\n","\n","        title = member.find('.//title').tag\n","        authors_head.append(title)\n","\n","        source = member.find('.//source').tag\n","        authors_head.append(source)\n","        \n","        pubmed = member.find('.//pubmedID').tag\n","        authors_head.append(pubmed)\n","       \n","        csvwriter.writerow(authors_head)\n","        count = count + 1\n","    \n","    author = member.find('.//authors').text\n","    authors.append(author) \n","    title = member.find('.//title').text\n","    authors.append(title)\n","    source = member.find('.//source').text\n","    authors.append(source)\n","    pubmed = member.find('.//pubmedID').text\n","    authors.append(pubmed)\n","         \n","    csvwriter.writerow(authors)\n"," \n"," \n","Ref_data.close()\n","\n","data= pd.read_csv(\"refdata.csv\")\n","data"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"8WBbLWHmWyBi"},"source":["## Reading the abstract from the first article in the list"]},{"cell_type":"code","metadata":{"id":"q4u48DIwUrvC"},"source":["from Bio import Entrez\n","# Searching for the abstract in the database Pubmed\n","Entrez.email = 'YOUR EMAIL HERE'\n","\n","query7 = Entrez.efetch(db='####', id = '####', rettype = '####', retmode = 'text')\n","# Hint for rettype (retrieval type): you want to retrieve the abstract\n","\n","# Reading the query and printing it\n","print(query7.read())\n","\n","# Closing the query\n","query7.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pfNB_qT2HF2Z"},"source":["## Answer the following questions:##\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"5fodSuNDa8r7"},"source":["1. Who is the first author on this article and what journal was it published in?\n"]},{"cell_type":"markdown","metadata":{"id":"EZRimSL8a_I9"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"keQrS6rdbAeJ"},"source":["2. Describe who was involved in the study (how many and what categories of patients?)."]},{"cell_type":"markdown","metadata":{"id":"8lBvaAtMbHtH"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"iLAuRr5YbKw9"},"source":["3. What did the researchers find out about K-Ras mutations?\n"]},{"cell_type":"markdown","metadata":{"id":"663GkS8AbPCB"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"UvUdY14qbR8Y"},"source":["4. What conclusion(s) did the researchers come to about K-Ras mutations based on their data? (Summarize and put into your own words)"]},{"cell_type":"markdown","metadata":{"id":"AR0UtYdybWBZ"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"Imzh1b3otgyk"},"source":["### Get information about genes related to the disease by setting 'geneMap' to 'TRUE'"]},{"cell_type":"code","metadata":{"id":"QxGJ-ivwMOoQ"},"source":["%%R\n","set_key('####')\n","\n","# Using mim number of 'LUNG CANCER' we search OMIM, setting 'geneMap' to TRUE\n","omim_result2 <- get_omim(####, geneMap = ####)\n","\n","# Save the results as an XML file\n","saveXML(omim_result2, file='FILE NAME HERE.xml\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oukCTVenzEFV"},"source":["### Display the results in the form of a table"]},{"cell_type":"code","metadata":{"id":"r-UtUSzxR0ny","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data2 = open('refdata2.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data2)\n","genes_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//phenotypeMap'):\n","    genes = []\n","    ref_list = []\n","    if count == 0:\n","        mim = member.find('.//mimNumber').tag\n","        genes_head.append(mim)\n","\n","        gen = member.find('.//geneSymbols').tag\n","        genes_head.append(gen)\n","\n","        phen = member.find('.//phenotype').tag\n","        genes_head.append(phen)\n","\n","        csvwriter.writerow(genes_head)\n","        count = count + 1\n","    \n","    mim = member.find('.//mimNumber').text\n","    genes.append(mim) \n","    gen = member.find('.//geneSymbols').text\n","    genes.append(gen)\n","    phen = member.find('.//phenotype').text\n","    genes.append(phen)\n","\n","    csvwriter.writerow(genes)\n"," \n"," \n","Ref_data2.close()\n","\n","data2= pd.read_csv(\"refdata2.csv\")\n","data2"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"I-a-OTjYzxzp"},"source":["Write down the KRAS mim number below"]},{"cell_type":"markdown","metadata":{"id":"fTL4mXBYz35S"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"CCZeuFF7z8R-"},"source":["## Using OMIM to search for gene information\n","\n","### First, read the description for the gene entry"]},{"cell_type":"code","metadata":{"id":"VDHLH2A3ZzII"},"source":["# Search OMIM again but with the KRAS mim number to obtain description information\n","%%R\n","set_key('####')\n","\n","\n","# Using mim number to get the description, set 'text' argument to true\n","omim_result <- get_omim(####, text = ####)\n","\n","saveXML(omim_result, file=\"FILE NAME HERE.xml\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ehmgPVBZ0Kpb"},"source":["### Display the results in the form of a table\n","\n"]},{"cell_type":"code","metadata":{"id":"EB4kI2S4btC8","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data3 = open('refdata3.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data3)\n","kras_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//entry'):\n","    kras = []\n","    ref_list = []\n","    if count == 0:\n","        des = member.find('.//textSectionList//textSectionContent').tag\n","        kras_head.append(des)\n","\n","        csvwriter.writerow(kras_head)\n","        count = count + 1\n","    \n","    des = member.find('.//textSectionList//textSectionContent').text\n","    kras.append(des) \n","\n","    csvwriter.writerow(kras)\n"," \n"," \n","Ref_data3.close()\n","\n","data3= pd.read_csv(\"refdata3.csv\")\n","pd.set_option('display.max_colwidth',1000)\n","\n","data3"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"-jjAPZqz0rKB"},"source":["### Now read information about allelic variants \n","An allele is a variant of a gene were the DNA sequence differs between two or more variants. \n","\n","Allelic variation describes the presence or number of different allele forms at a particular locus (locus or loci = place) on a chromosome.\n","\n","REF:  \n","* https://warwick.ac.uk/fac/sci/lifesci/research/vegin/geneticimprovement/diversitycollection/allelicvariation/\n"]},{"cell_type":"code","metadata":{"id":"xJqhcDReyBPC"},"source":["# Search OMIM again but with KRAS mim Number to obtain description info\n","%%R\n","set_key('####')\n","\n","# Set allelicVariantList to TRUE\n","omim_result <- get_omim(###, #### = ###)\n","\n","saveXML(omim_result, file='FILE NAME HERE.xml')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2Wm9udQGd6If"},"source":["### Display the results in the form of a table"]},{"cell_type":"code","metadata":{"id":"Fos9xqDVN1KQ","cellView":"form"},"source":["#@title Load the results by providing the file name in this form (include file extension .xml)\n","\n","# MAKING RESULTS LOOK GOOD\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","file_name = \"\" #@param {type:\"string\"}\n","\n","tree = ET.parse(file_name)\n","root = tree.getroot()\n"," \n","Ref_data4 = open('refdata4.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data4)\n","allele_head = []\n","\n"," \n","count = 0\n","for member in root.findall('.//allelicVariant'):\n","    allele = []\n","    ref_list = []\n","\n","    if count == 0:\n","      des = member.find('.//mutations').tag\n","      allele_head.append(des)\n","      \n","      mut = member.find('.//text').tag\n","      allele_head.append(mut)\n","     \n","      csvwriter.writerow(allele_head)\n","      count = count + 1\n","       \n","    des = member.find('.//mutations').text\n","    allele.append(des)\n","    \n","    mut = member.find('.//text').text\n","    allele.append(mut)\n","\n","    csvwriter.writerow(allele)\n","  \n","Ref_data4.close()\n","\n","data4= pd.read_csv(\"refdata4.csv\")\n","pd.set_option('display.max_colwidth',10000)\n","\n","data4\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"J0XqkjVAeUa9"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"iTIm0QiSef5J"},"source":["1. In the results above, read the entry for the Gly12Cys mutation. Describe the differences and similarities for the K-Ras, H-Ras, and N-Ras genes and proteins."]},{"cell_type":"markdown","metadata":{"id":"asWjhVSbekgV"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"cChziHozemJz"},"source":["2. How common was the G12C mutation in the Ahrendt et al. (2001) study?"]},{"cell_type":"markdown","metadata":{"id":"jWipNfQaeqZm"},"source":["Answer here"]}]}