{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #3: NCBI Gene and ExPASy Colab.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyMhjB8pSqBm2EXhZmzu9Vfm"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422–1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"097bbcf2"},"source":["# Translating the patient's cDNA sequence\n","\n","In order to investigate your mutated protein of interest, you need to know its amino acid sequence. Your patient comes in and you take a DNA sample. You send it in for sequencing and obtain its nucleotide sequence. This sequence is sotred in the following text file available to you via Google Drive: kRAS_PATIENT_cDNA.txt."]},{"cell_type":"markdown","metadata":{"id":"803dc301"},"source":["STEP #1: Open the cDNA sequence file in Python, read it and slightly edit it. By default, the text file contains some unformatted hidden characters. These hidden characters such as “/n” or “/r” needs to be formatted and removed.\n","Use the replace() function to edit the cDNA sequence txt file."]},{"cell_type":"code","metadata":{"id":"933278bc"},"source":["# Go to the files on Google Colab and right-click on the kRAS_PATIENT_cDNA file.\n","# Select 'Copy path' and paste it in the code below.\n","\n","# Assign the cDNA file a variable name (input_file).\n","input_file ='CODE HERE' \n","\n","# Open the input file:\n","f = open(input_file, 'r')\n","\n","# Read the file\n","seq = f.read()\n","\n","seq = seq.replace('\\n', '') # This replaces the new line character: '/n', with nothing: '', eliminating it.\n","seq = seq.replace('\\r', '') # This replaces the carriage return character (ENTER key character): '\\r', with nothing: '', eliminating it."],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"68bf9e43"},"source":["STEP #2: Build a function called translate() which will convert the cDNA sequence into its protein equivalent and return it. You will feed the edited cDNA sequence from the previous step as a parameter to the function. \n","\n","REF:\n","* https://www.geeksforgeeks.org/dna-protein-python-3/"]},{"cell_type":"code","metadata":{"id":"a9cecd15"},"source":["def translate(seq):\n","\n","# For the missing letters (???) next to the codons, add the corresponding one-letter code for the amino acids they code for.\n","\n","    table = {\n","        'ATA':'I', 'ATC':'I', 'ATT':'???', 'ATG':'M',\n","        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n","        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n","        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n","        'CTA':'L', 'CTC':'???', 'CTG':'L', 'CTT':'L',\n","        'CCA':'???', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n","        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n","        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n","        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n","        'GCA':'???', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n","        'GAC':'D', 'GAT':'???', 'GAA':'E', 'GAG':'E',\n","        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n","        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n","        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'???',\n","        'TAC':'???', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n","        'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n","    } \n","    \n","    protein ='' \n","\n","    for i in range(1, len(seq), 3): # reading frame 2 is determined by setting the range begin at 1. Here we will read our sequence in chunks of 3.\n","        codon = seq[i:i + 3] # Here we state that for every chunk of 3 amino acids, a codon is formed\n","\n","        if len(codon) == 3: # If the length of the codon is 3...\n","            protein += table[codon] # Match the codon to the corresponding amino acid in the table and form the protein sequence\n","            \n","    return protein # Our function returns the sequence"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"76dbdf14"},"source":["STEP #3: Use the translate() function to convert the cDNA sequence into its protein equivalent and return it. You will be using this sequence later on in the project. If you click on the three dots at the end, you can see the full sequence."]},{"cell_type":"code","metadata":{"id":"7ed9e518"},"source":["# The variable seq was set in a previous cell and contains our cDNA sequence. Here we use our translate function to obtain the sequence. \n","# Write seq inside the parenthesis and run the cell.\n","\n","translate(###)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0eHPSwferqay"},"source":["## Save your protein sequence\n"]},{"cell_type":"markdown","metadata":{"id":"kA-RndJpruNO"},"source":["You should be able to find the sequence of your protein by finding the first\n","methionine (M), then continuing until you see the first “*” which is a stop\n","codon. Copy the protein sequence in that region, starting with the first “M”\n","and paste it into the code below to save a document of your mutant protein sequence. Paste the sequence into the parenthesis after 'write'."]},{"cell_type":"code","metadata":{"id":"Cjj_isA8rpRU"},"source":["# Open the file you want to 'write' (w) into\n","text_file = open('mutant_protein.txt', 'w')\n","\n","# Write the string into the file\n","text_file.write('SEQUENCE HERE')\n","\n","# Close the file\n","text_file.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#IMPORTANT\n","\n","**Go to the folder at the right of the screen and download your mutant_protein.txt file. Upload it to your project_files folder.**"],"metadata":{"id":"GOBaBm_chOoL"}},{"cell_type":"markdown","metadata":{"id":"59e65969"},"source":["# Obtaining information about your gene of interest\n","\n","In this section, you will use web-scrapping to search the NCBI Gene database to obtain information about your gene of interest.\n","\n","## NCBI – Gene\n","**NCBI** – National Center for Biotechnology Information – This center was formed in 1988 as a division of the NLM (National Library of Medicine) at the NIH (National Institute of Health). As part of the NIH, NCBI is funded by the US government. The main goal of the center is to provide resources for biomedical researchers as well as the general public. The center is continually developing new materials and updating databases. The entire human genome is freely available on this website and is updated daily as new and better data becomes available. The NCBI homepage:http://www.ncbi.nlm.nih.gov.\n","\n","**Gene** is an NCBI database of genetic loci. This database used to be called LocusLink. Entries provide links to RefSeqs, articles in PubMed, and other descriptive information about genetic loci. The database also provides information on official nomenclature, aliases, sequence accession numbers, phenotypes, EC numbers, OMIM numbers, UniGene clusters, map information,\n","and relevant web sites.\n","\n","REF: \n","* Gene [Internet]. Bethesda (MD): National Library of Medicine (US), National Center for Biotechnology Information; 2004 – [cited 2021 Oct 31]. Available from: https://www.ncbi.nlm.nih.gov/gene/\n","* https://rsh249.github.io/python_workshop/scripting_NCBI_searches.html\n","* https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch"]},{"cell_type":"markdown","metadata":{"id":"hXev7oHiJMHM"},"source":["## Import the necessary packages from Biopython: \n","\n","**Entrez** is a molecular biology database system that provides integrated access to nucleotide and protein sequence data, gene-centered and genomic mapping information, 3D structure data, PubMed MEDLINE, and more. The system is produced by the National Center for Biotechnology Information (NCBI) and is available via the Internet.\n","\n","**SeqIO** provides a simple uniform interface to input and output assorted sequence file formats (including multiple sequence alignments), but will only deal with sequences as SeqRecord objects. \n","\n","REF:\n","* https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html\n","* https://biopython.org/wiki/SeqIO\n","\n"]},{"cell_type":"code","metadata":{"id":"5d5d79d0"},"source":["from Bio import Entrez\n","from Bio import SeqIO"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OeCNnjLDJSot"},"source":["## Begin by obtaining the gene's identification number from the NCBI.\n"]},{"cell_type":"markdown","metadata":{"id":"Ls1_1GUkCCTf"},"source":["To make use of NCBI's E-utilities, NCBI requires you to specify your\n","email address with each request.  As an example, if your email address\n","is A.N.Other@example.com, you can specify it as follows:\n","   from Bio import Entrez\n","   Entrez.email = 'A.N.Other@example.com'\n","In case of excessive usage of the E-utilities, NCBI will attempt to contact\n","a user at the email address provided before blocking access to the\n","E-utilities."]},{"cell_type":"code","metadata":{"id":"0r2R6LvjyT5X"},"source":["# Searching for the gene id in the database Gene\n","Entrez.email = 'YOUR EMAIL HERE'\n","\n","# Write KRAS+proto+oncogene+[homo sapiens(human)] in the space after 'term'.\n","\n","# db = database, in this case we are using the Gene database. Write gene after db.\n","query = Entrez.esearch(db='###', term='###')\n","\n","# Reading the query\n","record = Entrez.read(query)\n","\n","# Printing the 'IdList' from the record\n","record['IdList']"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"UBPBxo1np7hZ"},"source":["# Searching the database for a given id (write the id in the space after 'id') and choose the human protein entry\n","\n","# rettype - retrieval type, 'gb' = from GenBank, the NIH genetic sequence database\n","# retmode - retrieval mode, 'text' = results will display in the form of plain text\n","\n","query2 = Entrez.efetch(db ='###', id='###', rettype='gb', retmode='text')\n","\n","# Reading the query and printing it\n","print(query2.read())\n","\n","# Closing the query\n","query2.close()\n","\n","# If more than one ID came up, choose the homo sapiens one"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"6bc7e743"},"source":["## Fill in the following information from the Gene entry:\n","\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"a4febb69"},"source":["a. GeneID number:"]},{"cell_type":"markdown","metadata":{"id":"ef4ccafa"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"37298cd5"},"source":["b. Gene name:"]},{"cell_type":"markdown","metadata":{"id":"6fa87211"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"4fb3f535"},"source":["c. Where in the human genome is this gene located?"]},{"cell_type":"markdown","metadata":{"id":"c2d6c8ea"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"544c66d8"},"source":["# Obtaining information about your protein of interest\n","\n","In this section, you will use the Swiss-Prot packagr in BioPython to search the UniprotKB/Swiss-Prot database to obtain information about the protein product of your gene of interest."]},{"cell_type":"markdown","metadata":{"id":"fd2b8056"},"source":["## Import the necessary packages from BioPython and Python:\n","\n","**UniProtKB/Swiss-Prot** is a manually annotated, non-redundant protein sequence database. It combines information extracted from scientific literature and biocurator-evaluated computational analysis. The aim of UniProtKB/Swiss-Prot is to provide all known relevant information about a particular protein. Annotation is regularly reviewed to keep up with current scientific findings. The manual annotation of an entry involves detailed analysis of the protein sequence and of the scientific literature.\n","\n","The **ExPASy** (the Expert Protein Analysis System) World Wide Web server (http://www.expasy.org), is provided as a service to the life science community by a multidisciplinary team at the Swiss Institute of Bioinformatics (SIB). It provides access to a variety of databases and analytical tools dedicated to proteins and proteomics. \n","\n","Other Python modules:\n","\n","The **re** module implements regular expressions in Python. Regular expressions are an important tool for working with text carefully and effciently, they help to search for content in a string.\n","\n","**Requests** is a simple HTTP library. It allows you to send HTTP/1.1 requests easily.\n","\n","REF:\n","* https://en.wikipedia.org/wiki/UniProt\n","* Gasteiger, E., Gattiker, A., Hoogland, C., Ivanyi, I., Appel, R. D., & Bairoch, A. (2003). ExPASy: The proteomics server for in-depth protein knowledge and analysis. Nucleic acids research, 31(13), 3784–3788. https://doi.org/10.1093/nar/gkg563\n","* https://acme.byu.edu/00000179-af25-d5e1-a97b-bf6512cb0000/regex2020-pdf\n","* https://biopython-tutorial.readthedocs.io/en/latest/notebooks/10%20-%20Swiss-Prot%20and%20ExPASy.html#Parsing-Swiss-Prot-files\n","* https://biopython.org/docs/1.76/api/Bio.SwissProt.html\n","* https://web.expasy.org/docs/userman.html\n","* https://pypi.org/project/requests/"]},{"cell_type":"code","metadata":{"id":"eb1d5a96"},"source":["from Bio import ExPASy\n","from Bio import SwissProt\n","\n","import re\n","import requests"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"45f2e7bf"},"source":["STEP #1: Obtain the accession number of your protein"]},{"cell_type":"code","metadata":{"id":"8LxmcSWQ2pCf"},"source":["# Add name of protein and species to the 'full_URL' : GTPase+KRas and homo+sapiens\n","full_URL = ('http://www.uniprot.org/uniprot/?query=name%3A%22----CODE HERE----%22+AND+taxonomy%3A----CODE HERE----+AND+reviewed%3Ayes&format=list')\n","\n","# Obtain your results. Place 'full_URL' in the parenthesis after 'get'.\n","result = requests.get(###)\n","\n","# If the query is sucessful, print results, otherwise print error message and status code.\n","if result.ok:\n","  print(result.text)\n","else:\n","  print('Something went wrong', result.status_code)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"f720f935"},"source":["STEP #2: With the accession number of your protein, access information programatically using the Record Object's attributes. Just replicate the code below by changing attribute.\n","\n","record.attribute('accession number')\n","\n","#### Attributes:\n"," - entry_name - Name of this entry, e.g. RL1_ECOLI.\n"," - data_class - Either 'STANDARD' or 'PRELIMINARY'.\n"," - molecule_type - Type of molecule, 'PRT',\n"," - sequence_length - Number of residues.\n"," - accessions - List of the accession numbers, e.g. ['P00321']\n"," - created - A tuple of (date, release).\n"," - sequence_update - A tuple of (date, release).\n"," - annotation_update - A tuple of (date, release).\n"," - description - Free-format description.\n"," - gene_name - Gene name.  See userman.txt for description.\n"," - organism - The source of the sequence.\n"," - organelle - The origin of the sequence.\n"," - organism_classification - The taxonomy classification.  List of strings.\n","   (http://www.ncbi.nlm.nih.gov/Taxonomy/)\n"," - taxonomy_id - A list of NCBI taxonomy id's.\n"," - host_organism - A list of names of the hosts of a virus, if any.\n"," - host_taxonomy_id - A list of NCBI taxonomy id's of the hosts, if any.\n"," - references - List of Reference objects.\n"," - comments - List of strings.\n"," - cross_references - List of tuples (db, id1[, id2][, id3]).  See the docs.\n"," - keywords - List of the keywords.\n"," - features - List of tuples (key name, from, to, description).\n","   from and to can be either integers for the residue\n","   numbers, '<', '>', or '?'\n"," - protein_existence - Numerical value describing the evidence for the existence of the protein.\n"," - seqinfo - tuple of (length, molecular weight, CRC32 value)\n"," - sequence - The sequence."]},{"cell_type":"code","metadata":{"id":"b2b0cba4"},"source":["# Perform the query\n","query3 = ExPASy.get_sprot_raw('ACCESSION NUMBER HERE')\n","\n","# Read the results of the query\n","record = SwissProt.read(query3)\n","\n","# Assign the results a variable\n","res = record.description\n","\n","# Split description by the semicolon. This creates a list.\n","res2 = res.split(';')\n","\n","# Print the list with each item on a new line\n","print(*res2, sep ='\\n')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"3a6f9aa2"},"source":["# Print the sequence using the attribute 'sequence'\n","print(record.sequence)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"vEfATEO16Bn_"},"source":["# Print the comments in the entry using the attribute 'comments'\n","\n","# Assing the comments the variable 'res3'\n","res3 = record.###\n","\n","# Print the list with each item on a new line\n","print(*res3, sep = \"\\n\\n\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"zCfSfBAU0QSJ"},"source":["# Now access the features\n","\n","# YOUR CODE HERE"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"adf71978"},"source":["## Fill in the following information based on your ExPASy search:\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"91943a8a"},"source":["1. What tissues express this protein (under “comments”)?"]},{"cell_type":"markdown","metadata":{"id":"4715a002"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"c75678a8"},"source":["2. What is the chemical reaction catalyzed by this enzyme?\n"]},{"cell_type":"markdown","metadata":{"id":"ab0eee05"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"dfb3878b"},"source":["3. Where in the cell is this protein found?"]},{"cell_type":"markdown","metadata":{"id":"078f224a"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"3fc77f5f"},"source":["4. Which amino acid residue number binds the iron of the heme (under “Features”)?"]},{"cell_type":"markdown","metadata":{"id":"24543187"},"source":["Answer here"]}]}