{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #4_ NCBI BLAST and MSA Colab_phen.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyM/ZX17Wt0lQD0EWr5Gpidm"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422–1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"WuwAqUSQ1B2m"},"source":["# Comparing your mutant and wildtype protein to proteins in other species.\n","\n","In this section, you will use BioPython to perform Multiple Sequence Alignment of your mutated protein against other versions of the protein."]},{"cell_type":"markdown","metadata":{"id":"XG-FlbxYWFn5"},"source":["## BLAST and Multiple Sequence Alignment\n","\n","**BLAST** – Basic Local Alignment Search Tool is a program that compares a\n","sequence (input) to all the sequences in a database (that you choose). This\n","program aligns the most similar segments between sequences. BLAST aligns\n","sequences using a scoring matrix similar to BLOSUM (see entry). This scoring\n","method gives penalties for gaps and gives the highest score for identical\n","residues. Substitutions are scored based on how conservative the changes are.\n","The output shows a list of sequences, with the highest scoring sequence at the\n","top. The scoring output is given as an E-value. The lower the E-value, the\n","higher scoring the sequence is. E-values in the range of 1^-100 to 1^-50 are very similar (or even identical) sequences. Sequences with E-values 1^-10 and higher need to be examined based on other methods to determine homology. An Evalue of 1^-10 for a sequence can be interpreted as, “a 1 in 1^10 chance that the sequence was pulled from the database by chance alone (has no homology to the query sequence).” This program can be accessed from the NCBI homepage or:\n","http://www.ncbi.nlm.nih.gov/BLAST\n","\n","**ClustalW** is a program for making multiple sequence alignments. http://www.ebi.ac.uk/clustalw/index.html\n","\n","FASTA is a way of formatting a nuleic acid or protein sequence. It is important\n","because many bioinformatics programs require that the sequence be in FASTA\n","format. The FASTA format has a title line for each sequence that begins\n","with a “>” followed by any needed text to name the sequence. The end of\n","the title line is signified by a paragraph mark (hit the return key).\n","Bioinformatics programs will know that the title line isn’t part of the sequence if you have it formatted correctly. The sequence itself does NOT have any returns, spaces, or formatting of any kind. The sequence is given in one-letter code. An example of a protein in correct FASTA format is shown below:\n","\n",">K-Ras protein Homo sapiens\n","MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVIDGETCLLDI\n","LDTAGQEEYSAMRDQYMRTGEGFLCVFAINNTKSFEDIHHYREQIKRVKDSEDVP\n","MVLVGNKCDLPSRTVDTKQAQDLARSYGIPFIETSAKTRQGVDDAFYTLVREIRK\n","HKEKMSKDGKKKKKKSKTKCVIM\n","\n","\n","REF:\n","* Altschul, S.F., Gish, W., Miller, W., Myers, E.W. & Lipman, D.J.\n","(1990) \"Basic local alignment search tool.\" J. Mol. Biol. 215:403-410.\n","* W. R. Pearson (1990) “Rapid and Sensitive Sequence Comparison with FASTP\n","and FASTA” Methods in Enzymology 183:63-98."]},{"cell_type":"markdown","metadata":{"id":"CB2AIoip4U5l"},"source":["## Install and import the necessary packages from BioPython:\n","\n","The **Bio.Blast** package in BioPython has code for dealing with BLAST programs and output.\n","\n","**NCBIWWW** is a module from the Bio.Blast package of BioPython that provides code to work with the WWW version of BLAST provided by the NCBI. https://blast.ncbi.nlm.nih.gov/.\n","\n","**NCBIXML** is a module from the Bio.Blast package of BioPython that provides code to work with the BLAST XML output file.\n","\n","**MAFFT** (Multiple Alignment using Fast Fourier Transform) is a multiple sequence alignment program.\n","\n","**Bio.Align.Applications** is a BioPython package that contains the MafftCommandline module which allows us to use MAFFT to align our sequences.\n","\n","**Bio.AlignIO** is a BioPython package that reads and writes multiple sequence alignment files (generated by mafft) \n","\n","**Entrez** is a molecular biology database system that provides integrated access to nucleotide and protein sequence data, gene-centered and genomic mapping information, 3D structure data, PubMed MEDLINE, and more. The system is produced by the National Center for Biotechnology Information (NCBI) and is available via the Internet.\n","\n","**SeqIO** provides a simple uniform interface to input and output assorted sequence file formats (including multiple sequence alignments), but will only deal with sequences as SeqRecord objects. \n","\n","REF:\n","* https://biopython.org/docs/1.75/api/Bio.Blast.html\n","* https://biopython.org/docs/1.75/api/Bio.Blast.NCBIWWW.html\n","* https://biopython.org/docs/1.76/api/Bio.Blast.NCBIXML.html\n","* https://biopython.org/docs/1.75/api/Bio.Align.Applications.html\n","* https://biopython.org/wiki/AlignIO\n","* https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html\n","* https://biopython.org/wiki/SeqIO"]},{"cell_type":"code","metadata":{"id":"aSIKennH461M"},"source":["from Bio.Blast import NCBIWWW\n","from Bio.Blast import NCBIXML\n","!apt-get install -qq -y mafft\n","from Bio.Align.Applications import MafftCommandline\n","from Bio import AlignIO\n","from Bio import Entrez\n","from Bio import SeqIO"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UPtS1D-_X1HN"},"source":["STEP #1: Obtain FASTA sequences of proteins homologus to your mutated protein by performing a BLAST search with the unmutated sequence obtained from NCBI gene.\n","\n","REF:\n","* https://hbr.org/2000/07/explaining-xml#:~:text=XML%20stands%20for%20extensible%20markup,used%20to%20format%20Web%20pages. \n"]},{"cell_type":"code","metadata":{"id":"TSt02NZRzFyQ"},"source":["# Perform the query. \n","\n","# Paste your protein sequence from Uniprot in Notebook #3, inside query4 where it says SEQUENCE HERE.\n","\n","# format_type - format of results, in this case we want 'XML'.\n","\n","# XML (extensible markup language), similar to html. A markup language is a set of codes, or tags, that describes the text in a digital document.\n","query4 = NCBIWWW.qblast('blastp','nr','SEQUENCE HERE', format_type = 'FORMAT HERE')\n","\n","# Open the file you will 'write' (w) into\n","res4 = open('NAME THIS FILE.xml', 'w')\n","\n","# Write the file with the query results\n","res4.write(query4.read())\n","\n","# Close the file\n","res4.close()\n","\n","# Close the query\n","query4.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2IGdSN8r12lw"},"source":["STEP #2: Inspect file with BLAST results. Observe all the sequences that aligned with thesequence of your protein of interest."]},{"cell_type":"code","metadata":{"id":"IZePx_-G7qdJ","cellView":"form"},"source":["#@title Inspecting the BLAST results file. Write the name (including file extension) of the file with your BLAST results. { run: \"auto\", vertical-output: true }\n","\n","# This code is for making the results look good. It will be hidden to the students.\n","import xml.etree.ElementTree as ET\n","import csv\n","import pandas as pd\n","\n","File_with_BLAST_Results = '' #@param {type:\"string\"}\n","\n","tree = ET.parse(File_with_BLAST_Results)\n","root = tree.getroot()\n","\n","Ref_data = open('blast.csv', 'w')\n"," \n","csvwriter = csv.writer(Ref_data)\n","main_head = []\n"," \n","count = 0\n","for member in root.findall('.//Hit'):\n","    main = []\n","    ref_list = []\n","    if count == 0:\n","        num = member.find('.//Hit_num').tag\n","        main_head.append(num)\n","\n","        id = member.find('.//Hit_id').tag\n","        main_head.append(id)\n","        \n","        acc = member.find('.//Hit_accession').tag\n","        main_head.append(acc)\n","\n","        lenght = member.find('.//Hit_len').tag\n","        main_head.append(lenght)\n","\n","        hitseq = member.find('.//Hsp_hseq').tag\n","        main_head.append(hitseq)\n","       \n","        csvwriter.writerow(main_head)\n","        count = count + 1\n","\n","    num = member.find('.//Hit_num').text\n","    main.append(num)\n","\n","    id = member.find('.//Hit_id').text\n","    main.append(id)\n","        \n","    acc = member.find('.//Hit_accession').text\n","    main.append(acc)\n","\n","    lenght = member.find('.//Hit_len').text\n","    main.append(lenght)\n","\n","    hitseq = member.find('.//Hsp_hseq').text\n","    main.append(hitseq)\n","             \n","    csvwriter.writerow(main)\n","  \n","Ref_data.close()\n","\n","data= pd.read_csv(\"blast.csv\")\n","pd.set_option('display.max_colwidth',10000)\n","\n","data"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"nu2lHAsAcfd0"},"source":["## Obtaining a list of all the organisms whose sequences aligned with your protein of interest."]},{"cell_type":"code","metadata":{"id":"_Xzmy8qsynMs"},"source":["# Open the file with your BLAST results. \n","blast_results = open('NAME OF FILE.xml')\n","\n","# Parse the file\n","blast_records = NCBIXML.parse(blast_results)\n","\n","for blast_rec in blast_records:\n","  for description in blast_rec.descriptions: # For every record in the file...\n","    results = description.title # ...extract the title from the description\n","\n","    div_res = results.split('>') # Divide the results whenever there is a '>'\n","    print(*div_res, sep = \"\\n\\n\") # Display results double spaced (two 'new lines', '\\n')\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"qAy8bm86lbPt"},"source":["After obtaining the results, choose 5 organisms to compare their protein sequences to that of your mutated protein sequence. The goal is to choose a variety of sequences that differ in evolutionary distance from the human protein. Write the 5 organisms below together with their reference number. Include at the end, the human sequence's reference number. It should be at the top of the list in the previous cell (after 'gb'). Later on, we will add your mutated protein sequence to the file for a total of 7 sequences."]},{"cell_type":"markdown","metadata":{"id":"uLH_2OrUoasp"},"source":["Answer here:\n","1. Organism 1\n","2. Organism 2\n","3. Organism 3\n","4. Organism 4\n","5. Organism 5\n","6. Human protein"]},{"cell_type":"markdown","metadata":{"id":"AvLCkoKmX4CX"},"source":["STEP #4: Obtain FASTA sequences for all 6 sequences above and make a file with all 6 sequences. Then add the mutated sequence to that file. This will be used for multiple sequence alignment. You will obtain the seqeunces from the NCBI Protein database."]},{"cell_type":"code","metadata":{"id":"dkY_hif5n9VU"},"source":["# Perform the query. Add the reference numbers after 'id'.\n","Entrez.email = 'EMAIL HERE' # You need to add your email again\n","\n","# Now the database is protein. Change it accordingly\n","query5 = Entrez.efetch(db='DATABASE HERE', id='IDS OF ORGANISMS HERE SEPARATED BY A COMMA', rettype='gb', retmode='text')\n","\n","# Open the file you want to write into. The file extension should be .fa (FASTA format)\n","unaligned_sequences = open('NAME THIS FILE.fa', 'w')\n","\n","entries = []\n","\n","for seq_record in SeqIO.parse(query5, 'gb'): # For every record in the query (which we parse with SeqIO)...\n","\n","  # ... print the record id (seq_record.id), source, sequence length, and sequence, \n","  # separated by spaces (' '), and in a new line each time ('\\n')...\n","  print(seq_record.id + ' ' + seq_record.annotations['source'] + ': ' + 'Sequence length = ' + str(len(seq_record)) + '\\n' + seq_record.seq + '\\n')\n","\n","  seq_record.id = seq_record.id + ' ' + seq_record.annotations['source']\n","\n","  # ... use SeqIO to write each record (seq_record) to our file (unaligned_sequences), in FASTA format.\n","  SeqIO.write(seq_record, unaligned_sequences, 'fasta')\n","\n","  entries.append(seq_record.annotations['source'])\n","\n","# Close the file\n","unaligned_sequences.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Add the mutated patient sequence to the file you created in the previous cell.\n","To do this, you must first upload your mutant_protein.txt file to the local drive space.\n"],"metadata":{"id":"ix87vlOb7eqa"}},{"cell_type":"code","metadata":{"id":"djQVofvNmIgk"},"source":["# entering the file names\n","firstfile = 'NAME OF FILE.fa'\n","secondfile = 'mutant_protein.txt'\n"," \n","# opening both files in read only mode to read initial contents\n","f1 = open(firstfile, 'r')\n","f2 = open(secondfile, 'r')\n"," \n","# printing the contens of the file before appending\n","print('content of first file before appending -', f1.read())\n","print('content of second file before appending -', f2.read())\n"," \n","# closing the files\n","f1.close()\n","f2.close()\n"," \n","# opening first file in append mode and second file in read mode\n","f1 = open(firstfile, 'a+')\n","f2 = open(secondfile, 'r')\n"," \n","# appending the contents of the second file to the first file\n","f1.write('>mutated_sequence_patient' + '\\n' + f2.read())\n","\n","# relocating the cursor of the files at the beginning\n","f1.seek(0)\n","f2.seek(0)\n"," \n","# printing the contents of the files after appendng\n","print('content of first file after appending -', f1.read())\n","print('content of second file after appending -', f2.read())\n"," \n","# closing the files\n","f1.close()\n","f2.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0oHwpwYTFDvY"},"source":["STEP #5: Perform Multiple Sequence Alignment."]},{"cell_type":"code","metadata":{"id":"J77hi9Hb6Gsc"},"source":["# Initialize the program with the FASTA file\n","mafft_cline = MafftCommandline(input='NAME OF FILE.fa')\n","\n","# Set the outputs of maft_cline() to go into stdout and stderr\n","stdout, stderr = mafft_cline()\n","\n","# Open a file to write into called 'aligned.fasta'... \n","with open('NAME HERE.fasta', 'w') as handle:\n","  handle.write(stdout) # ...and write the output of the alignment to stdout and into aligned.fasta\n","\n","# Read the alignment file using AlignIO\n","align = AlignIO.read('NAME OF FILE.fasta', 'fasta')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KwE5gXTXUwyq"},"source":["## View your Multiple Sequence Alignment File\n","\n","In the form below, write the name of the file which has the alignment and its format.\n"]},{"cell_type":"code","metadata":{"id":"EoRs5EUl6j4s","cellView":"form"},"source":["#@title Load the viewer by indicating the MSA file (including file extension .txt) and format to read  { vertical-output: true }\n","#The following code is modified from the wonderful viewer developed by Damien Farrell\n","#https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n","\n","#Importing all modules first\n","import os, io, random\n","import string\n","import numpy as np\n","\n","from Bio.Seq import Seq\n","from Bio.Align import MultipleSeqAlignment\n","from Bio import AlignIO, SeqIO\n","\n","import panel as pn\n","import panel.widgets as pnw\n","pn.extension()\n","\n","from bokeh.plotting import figure\n","from bokeh.models import ColumnDataSource, Plot, Grid, Range1d\n","from bokeh.models.glyphs import Text, Rect\n","from bokeh.layouts import gridplot\n","\n","#Setting up the amino color code according to Zappo color scheme\n","def get_colors(seqs):\n","    #make colors for bases in sequence\n","    text = [i for s in list(seqs) for i in s]\n","    #Use Zappo color scheme\n","    clrs =  {'K':'red',\n","             'R':'red',\n","             'H':'red',             \n","             'D':'green',\n","             'E':'green',\n","             'Q':'blue',\n","             'N':'blue',\n","             'S':'blue',\n","             'T':'blue',\n","             'A':'blue',\n","             'I':'blue',\n","             'L':'blue',\n","             'M':'blue',\n","             'V':'blue',\n","             'F':'orange',\n","             'Y':'orange',\n","             'W':'orange',\n","             'C':'blue',\n","             'P':'yellow',\n","             'G':'orange',\n","             '-':'white'}\n","    colors = [clrs[i] for i in text]\n","    return colors\n","\n","#Setting up the MSA viewer\n","def view_alignment(aln, fontsize=\"9pt\", plot_width=800):\n","    \"\"\"Bokeh sequence alignment view\"\"\"\n","\n","    #make sequence and id lists from the aln object\n","    seqs = [rec.seq for rec in (aln)]\n","    ids = [rec.id for rec in aln]    \n","    text = [i for s in list(seqs) for i in s]\n","    colors = get_colors(seqs)    \n","    N = len(seqs[0])\n","    S = len(seqs)    \n","    width = .4\n","\n","    x = np.arange(1,N+1)\n","    y = np.arange(0,S,1)\n","    #creates a 2D grid of coords from the 1D arrays\n","    xx, yy = np.meshgrid(x, y)\n","    #flattens the arrays\n","    gx = xx.ravel()\n","    gy = yy.flatten()\n","    #use recty for rect coords with an offset\n","    recty = gy+.5\n","    h= 1/S\n","    #now we can create the ColumnDataSource with all the arrays\n","    source = ColumnDataSource(dict(x=gx, y=gy, recty=recty, text=text, colors=colors))\n","    plot_height = len(seqs)*15+50\n","    x_range = Range1d(0,N+1, bounds='auto')\n","    if N>100:\n","        viewlen=100\n","    else:\n","        viewlen=N\n","    #view_range is for the close up view\n","    view_range = (0,viewlen)\n","    tools=\"xpan, xwheel_zoom, reset, save\"\n","\n","    #entire sequence view (no text, with zoom)\n","    p = figure(title=None, plot_width= plot_width, plot_height=50,\n","               x_range=x_range, y_range=(0,S), tools=tools,\n","               min_border=0, toolbar_location='below')\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                 line_color=None, fill_alpha=0.6)\n","    p.add_glyph(source, rects)\n","    p.yaxis.visible = False\n","    p.grid.visible = False  \n","\n","    #sequence text view with ability to scroll along x axis\n","    p1 = figure(title=None, plot_width=plot_width, plot_height=plot_height,\n","                x_range=view_range, y_range=ids, tools=\"xpan,reset\",\n","                min_border=0, toolbar_location='below')#, lod_factor=1)          \n","    glyph = Text(x=\"x\", y=\"y\", text=\"text\", text_align='center',text_color=\"black\",\n","                text_font=\"monospace\",text_font_size=fontsize)\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                line_color=None, fill_alpha=0.4)\n","    p1.add_glyph(source, glyph)\n","    p1.add_glyph(source, rects)\n","\n","    p1.grid.visible = False\n","    p1.xaxis.major_label_text_font_style = \"bold\"\n","    p1.yaxis.minor_tick_line_width = 0\n","    p1.yaxis.major_tick_line_width = 0\n","\n","    p = gridplot([[p],[p1]], toolbar_location='below')\n","    return p\n","\n","#Loading the viewer by indicating the MSA file and format to read\n","#@markdown \n","MSA_file = '' #@param {type:\"string\"}\n","MSA_format = '' #@param {type:\"string\"}\n","aln = AlignIO.read(MSA_file,MSA_format)\n","p = view_alignment(aln, plot_width=900)\n","pn.pane.Bokeh(p)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OPNy1UU-VHlQ"},"source":["Press the arrow below the alignment (Pan (x-axis)). This allows you to click on the alignment and draw from side to side to view the entire sequence alignment.\n"]},{"cell_type":"markdown","metadata":{"id":"XOk5obBVm8Vq"},"source":["## Answer the following questions based on your analysis of the Multiple Sequence Alignment results.\n","Input your answer in the cell below each question and press SHIFT+ENTER."]},{"cell_type":"markdown","metadata":{"id":"PrOUmf9VnJns"},"source":["1. What is the mutation in the mutant sequence? Write it in the following\n","format “Res123Res” where the first Res is the three-letter code for the\n","amino acid in the un-mutated (wild type) protein and the second Res is the\n","amino acid in the mutated protein. In place of “123” put the amino acid\n","residue number of the mutation."]},{"cell_type":"markdown","metadata":{"id":"jtE6LMB7XSXQ"},"source":["Answer here"]},{"cell_type":"markdown","source":["2. Describe how is conservation defined in this alignment diagram."],"metadata":{"id":"ZX1TkhOjwTpr"}},{"cell_type":"markdown","source":["Answer here"],"metadata":{"id":"VmFzd1c9wV0e"}},{"cell_type":"markdown","metadata":{"id":"BOfQuTHDWsm_"},"source":["3. Is your patient’s mutation in a region of conservation?\n"]},{"cell_type":"markdown","metadata":{"id":"OSMfTXBzXUd_"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"hnCDsIneWszl"},"source":["4. What is the secondary structure predicted for the region containing your patient’s mutation? Found in UniProt.\n"]},{"cell_type":"markdown","metadata":{"id":"2-yvNIqSXUXI"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"Ij81BFBeWs_f"},"source":["5. Based on the alignment, what span of amino acids is LEAST conserved?\n"]},{"cell_type":"markdown","metadata":{"id":"5rwhPEtIXZ8x"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"kFyLOut2WtPX"},"source":["6. What secondary structure is predicted in this area of low conservation? Found in UniProt."]},{"cell_type":"markdown","metadata":{"id":"fuL2pmh-WteM"},"source":["Answer here"]}]}