{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Notebook #8_ KEGG Colab_hmg.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyM8aT2jryN5Kuvcsf7AOkHn"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"e4_-znEWZ8AB"},"source":["First you need to link your Google Drive to the notebook in order to access the files needed for this module.\n","\n","Run the cell below and follow instructions to mount the drive."]},{"cell_type":"code","metadata":{"id":"SHOuF0YNHktr"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OrHpyyZhi0La"},"source":["## Installing Biopython\n","\n","At the beginning of each module, we will install **Biopython**. Biopython is a large open-source application programming interface (API) used in both bioinformatics software development and in everyday scripts for common bioinformatics tasks. It contains several packages that you will need to import which will allow you to run the analyses required for this project. \n","\n","REF:\n","* Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & de Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics (Oxford, England), 25(11), 1422â€“1423. https://doi.org/10.1093/bioinformatics/btp163\n"]},{"cell_type":"code","metadata":{"id":"7hOWkK_qiyt_"},"source":["!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zpxieJg51lIN"},"source":["## KEGG Search for Metabolic Pathways"]},{"cell_type":"markdown","metadata":{"id":"xoWr6Iwr3Q-V"},"source":["**KEGG** (Kyoto Encyclopedia of Genes and Genomes) is a website is used for\n","accessing metabolic pathways. At this website, you can search a process, gene, protein, or metabolite and obtain diagrams of all the metabolic pathways\n","associated with your query. \n","\n","REF:    \n","* http://www.genome.ad.jp/kegg/\n","* https://widdowquinn.github.io/2018-03-06-ibioic/02-sequence_databases/09-KEGG_programming.html\n"]},{"cell_type":"markdown","metadata":{"id":"_iKZ4Cpd4DXM"},"source":["## Install and import the necessary packages from BioPython, Python and IPython:\n","\n","**ReportLab** is an open-source library fpr generating PDFs and graphics.\n","\n","**Image** is a class that is part of the **display** submodule of **IPython** (interactive Python). It allows for images to be displayed in the notebook.\n","\n","**Bio.KEGG** is a BioPython module with code to work with data from the KEGG database. **REST** is a submodule that makes navigating KEGG easier.\n","\n","**Pandas** is a Python library that is used to analyze data. It will help create dataframes (tables of data) to store the data obtained from KEGG.\n","\n","**io** is a module part of the Python standard library that will allow usto manage files.\n","\n","REF:\n","* https://pypi.org/project/reportlab/\n","* https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html\n","* https://biopython.org/docs/dev/api/Bio.KEGG.html\n","* https://www.w3schools.com/python/pandas/default.asp\n","* https://www.journaldev.com/19178/python-io-bytesio-stringio"]},{"cell_type":"code","metadata":{"id":"C17M06h86Jhs"},"source":["!pip install reportlab"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GzWy7ww_3RqF"},"source":["# Show images inline\n","from IPython.display import Image\n","\n","# Import Biopython modules to interact with KEGG\n","from Bio.KEGG import REST\n","\n","# Import Pandas, so we can use dataframes\n","import pandas as pd\n","\n","# Import io to manage files\n","import io"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"i6_xK7rN578j"},"source":["# This code that will help us display the PDF output\n","def PDF(filename):\n","    return HTML('<iframe src=%s width=700 height=350></iframe>' % filename)\n","\n","# This code returns a Pandas dataframe, given tabular text\n","def to_df(result):\n","    return pd.read_table(io.StringIO(result), header=None)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"wN7KShsy7-us"},"source":["# Find a specific entry with a precise search term \n","\n","# Use the gene's identification number obtained in Notebook #3\n","\n","# Store results in a variable called result\n","result = REST.kegg_find('genes', '3156').read()\n","\n","# This just lets us see more of the columns\n","pd.set_option('display.max_colwidth',10000)\n","\n","# Display results as a datatframe (df)\n","to_df(result)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vrfzXq5s0ZmV"},"source":["Write down the code in the second column that corresponds to HMGCR below."]},{"cell_type":"markdown","metadata":{"id":"vcbgVMIh0jT4"},"source":["Answer here"]},{"cell_type":"code","metadata":{"id":"EjHArXXr6daM"},"source":["# Get the entry information\twith the ID in the previous cell\n","\n","# Store it in a variable called result2\n","#### = REST.kegg_get('####').read()\n","\n","# Display the results\n","print(####)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zsE8gSYYCVg8"},"source":["From PATHWAY above, select Terpenoid backbone biosynthesis and annotate the code before it:\n"]},{"cell_type":"markdown","metadata":{"id":"AQABjSVj0tAh"},"source":["Answer here"]},{"cell_type":"markdown","metadata":{"id":"PhriCO31SL2P"},"source":["In biochemistry, a metabolic pathway is a linked series of chemical reactions occurring within a cell. The reactants, products, and intermediates of an enzymatic reaction are known as metabolites, which are modified by a sequence of chemical reactions catalyzed by enzymes.\n","\n","In the next cell, you will obtain a chart of **Terpenoid backbone biosynthesis**. \n","\n","Terpenoids, also known as isoprenoids, are a large class of natural products consisting of isoprene (C5) units. There are two biosynthetic pathways, the mevalonate pathway and the non-mevalonate pathway or the MEP/DOXP pathway, for the terpenoid building blocks.: isopentenyl diphosphate (IPP) and dimethylallyl diphosphate (DMAPP). The action of prenyltransferases then generates higher-order building blocks: geranyl diphosphate (GPP), farsenyl diphosphate (FPP), and geranylgeranyl diphosphate (GGPP), which are the precursors of monoterpenoids (C10), sesquiterpenoids (C15), and diterpenoids (C20), respectively. Condensation of these building blocks gives rise to the precursors of sterols (C30) and carotenoids (C40). The MEP/DOXP pathway is absent in higher animals and fungi, but in green plants the MEP/DOXP and mevalonate pathways co-exist in separate cellular compartments. The MEP/DOXP pathway, operating in the plastids, is responsible for the formation of essential oil monoterpenes and linalyl acetate, some sesquiterpenes, diterpenes, and carotenoids and phytol. The mevalonate pathway, operating in the cytosol, gives rise to triterpenes, sterols, and most sesquiterpenes."]},{"cell_type":"code","metadata":{"id":"J_-CUy9E6sdE"},"source":["# Get map of Terpenoid backbone biosynthesis\n","\n","# Input the ID for the Terpenoid backbone biosynthesis preceded by hsa\n","# Store results in variable called result3\n","\n","#### = REST.kegg_get('ID HERE', 'image').read()\n","\n","# Display image\n","Image(####)\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"g4mJC8QVE6qK"},"source":["## Answer the following questions:\n","Input your answer in the cell below each question and press SHIFT+ENTER.\n"]},{"cell_type":"markdown","metadata":{"id":"2D34d_c-018f"},"source":["1. What metabolite is directly reduced by inhibiting HMG-CoA Reductase and what other pathways (besides cholesterol biosynthesis) does this affect?\n"," "]},{"cell_type":"markdown","metadata":{"id":"vlCzF8491AVK"},"source":["Answer here"]}]}